ls1intum/Artemis
# `Notifications`: Improve notification text translation

### Checklist
#### General


#### Server
#### Client

### Motivation and Context


Previously users only received English notifications even if they set the language to German. Now the notification language depends on the selected language like the rest of the client. The notification texts are also improved to give a more meaningful description.

### Description

Instead of sending English notification titles and texts to the client we now send a text placeholder. This placeholder is then replaced with the actual text in the client depending on the selected language. Notification texts that are entered as free text from an instructor stay untouched. To ensure that we introduced a new column to the notification table in the database with a flag that indicates whether or not the notification consists of a text placeholder. We also added a second column to the same table to store information that is later needed to construct the actual text in the client (e.g. course name or exercise name).
To be able to also handle notifications that were generated previous to these changes we set the database entries in the column 'is_placeholder' for old notifications to false. This way they are always treated like free texts from the client. This means they will not be translated even if it would be possible for this notification type, but we are still able to display them.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student

1. Log in to Artemis
2. Trigger some notifications (There are many notifications in Artemis so providing an exhaustive list is hard. There are many notifications when using the communication tab e.g. when posts get published or someone responds to a post or in tutorial groups e.g. if someone registers for the tutorial group or leaves it. Also when archiving an exam or a course, when exercises get released, when quizzes start, when there is a plagiarism case etc.)
3. You can look at all your received notifications if you click on the bell in the top right of the screen. Important notifications also come in a pop-up.
4. Switch language in between generating notifications. 
5. Verify that all notifications are generated and are always in the selected language

#### Exam Mode Testing



Prerequisites:
- 1 Instructor
- 2 Students
- 1 Exam with a Programming Exercise

1. Log in to Artemis
2. Participate in the exam as a student
3. Change an exercise during the exam as an instructor
4. Verify that you get notified as a student

### Review Progress






#### Code Review
#### Manual Tests
#### Exam Mode Test

### Test Coverage





| Class/File | Branch | Line |
|------------|-------:|-----:|
| notification-sidebar.component.ts| 65.34%| 83.67% |
| notification-popup.component.ts | 72.28% | 92.3% |
| notification.service.ts | 57% | 78.57% |
| SingleUserNotification.java | n/a| 100% |
| GroupNotification.java | n/a| 89% |
| TutorialGroupNotification.java | 50% | 85% |
| SingleUserNotificationFactory.java | 66% | 92% |
| GroupNotificationFactory.java | 85%| 94% |
| TutorialGroupNotificationFactory.java | 85% | 100% |
| SingleUserNotificationService.java | 94%| 98% |
| GroupNotificationService.java | 90% | 99% |
| NotificationSettingsService.java | 95% | 100% |

### Screenshots


English Notifications

<img width="299" alt="Bildschirm­foto 2023-03-29 um 12 21 01" src="https://user-images.githubusercontent.com/44734881/228504813-ed995aab-c454-48e3-9461-fdd869ffcc93.png">

German Notifications

<img width="299" alt="Bildschirm­foto 2023-03-29 um 12 21 22" src="https://user-images.githubusercontent.com/44734881/228504863-f80c3619-77dc-488e-aa5d-4bbb3f2824ad.png">
